### PR TITLE
First implementation for SiltBreaker Ogre Tank Boss 

### PR DESCRIPTION
**Pull request by carlosrpg**

### CODE REVIEW

**Comment by chrisinajar** [comment]
Use an editor that supports `.editorconfig` such as Sublime Text (with the extension installed) to prevent these trivial whitespace errors

File: game/scripts/vscripts/units/ai_ogre_tank_boss.lua [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	if thisEntity.bStarted == false then
+		return 0.1
+	else
+		thisEntity:RemoveModifierByName( "modifier_invulnerable" )
+		--print( "removed invuln modifier from ogre boss" )
+	end
+
+	-- Are we too far from our initial spawn position?
+	local fDist = ( thisEntity:GetOrigin() - thisEntity.vInitialSpawnPos ):Length2D()
+	if fDist > 1000 then
```
Comment (chrisinajar) [inline game/scripts/vscripts/units/ai_ogre_tank_boss.lua]: That's a really short leash, 2k would be more reasonable

**Comment by carlosrpg** [comment]
Finally!

File: game/scripts/vscripts/units/ai_ogre_tank_boss.lua [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		end
+	end
+
+	if thisEntity.JumpAbility ~= nil and thisEntity.JumpAbility:IsFullyCastable() and nEnemiesRemoved > 0 then
+		return Jump()
+	end
+
+	if #enemies == 0 then
+		-- @todo: Could check whether there are ogre magi nearby that I should be positioning myself next to.  Either that or have the magi come to me.
+		return 1
```
Comment (chrisinajar) [inline game/scripts/vscripts/units/ai_ogre_tank_boss.lua]: return home here, if there are no enemies the boss should just head back

File: game/scripts/vscripts/units/ai_ogre_tank_boss.lua [hunk 3]
```diff
@@ -0,0 +1,120 @@
+
```
Comment (chrisinajar) [inline game/scripts/vscripts/units/ai_ogre_tank_boss.lua]: You need to `LinkLuaModifier` the modifiers used in this file

File: game/scripts/vscripts/modifiers/modifier_ogre_tank_melee_smash_thinker.lua [hunk 4]
```diff
@@ -0,0 +1,76 @@
+
+modifier_ogre_tank_melee_smash_thinker = class({})
```
Comment (chrisinajar) [inline game/scripts/vscripts/modifiers/modifier_ogre_tank_melee_smash_thinker.lua]: Use the modifier base class here

File: game/scripts/vscripts/abilities/siltbreaker/npc_dota_creature_ogre_tank_boss/ogre_tank_boss_melee_smash.lua [hunk 5]
```diff
@@ -0,0 +1,35 @@
+ogre_tank_boss_melee_smash = class({})
```
Comment (chrisinajar) [inline game/scripts/vscripts/abilities/siltbreaker/npc_dota_creature_ogre_tank_boss/ogre_tank_boss_melee_smash.lua]: ability base class

File: game/scripts/vscripts/abilities/siltbreaker/npc_dota_creature_ogre_tank_boss/ogre_tank_boss_jump_smash.lua [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+-----------------------------------------------------------------------------
+
+function ogre_tank_boss_jump_smash:ProcsMagicStick()
+	return false
+end
+
+-----------------------------------------------------------------------------
+
+function ogre_tank_boss_jump_smash:GetPlaybackRateOverride()
+	return 0.9 -- keep this proportional to jump_speed
```
Comment (chrisinajar) [inline game/scripts/vscripts/abilities/siltbreaker/npc_dota_creature_ogre_tank_boss/ogre_tank_boss_jump_smash.lua]: should probably return `self:GetSpecialValueFor("jump_speed") / 2`

File: game/scripts/vscripts/abilities/siltbreaker/npc_dota_creature_ogre_tank_boss/ogre_tank_boss_jump_smash.lua [hunk 7]
```diff
@@ -0,0 +1,26 @@
+
+ogre_tank_boss_jump_smash = class({})
```
Comment (chrisinajar) [inline game/scripts/vscripts/abilities/siltbreaker/npc_dota_creature_ogre_tank_boss/ogre_tank_boss_jump_smash.lua]: use ability base class

File: game/scripts/vscripts/units/ai_ogre_tank_boss.lua [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	thisEntity:AddNewModifier( nil, nil, "modifier_invulnerable", { duration = -1 } )
+
+	thisEntity.SmashAbility = thisEntity:FindAbilityByName( "ogre_tank_boss_melee_smash" )
+	thisEntity.JumpAbility = thisEntity:FindAbilityByName( "ogre_tank_boss_jump_smash" )
+
+	thisEntity:SetContextThink( "OgreTankBossThink", OgreTankBossThink, 1 )
+end
+
+function OgreTankBossThink()
+	if ( not thisEntity:IsAlive() ) then
```
[the remark this answers is not in the document]
Comment (Trildar) [reply game/scripts/vscripts/units/ai_ogre_tank_boss.lua]: `if IsValidEntity(thisEntity) and not thisEntity:IsAlive() then` works as well I believe, and I think Gester/Keril recommended using `IsValidEntity` over `IsNull`.

**Comment by carlosrpg** [comment]
Code Review implemented. build is already failing, not from my code

**Comment by chrisinajar** [comment]
When I test this locally the boss is a giant wireframe. What do i need to compile?

**Comment by chrisinajar** [comment]
"Cannot recompile: source file not present" -- I think you need to run `git add -A` and then commit again?

**Comment by chrisinajar** [comment]
Yeah I think i'm missing the files in the content dir? The UI also shows "Content File: (no file)"

**Comment by carlosrpg** [comment]
Thats weird. I staged and commited everything.
I just copied the model/creeps/ogre_1 folder. it is in the first commit.
Also, I did not compiled it. I thought I was copying the compiled version _c already

**Comment by warpdragon** [comment]
how do you make a boss without editing a single source file?

**Comment by carlosrpg** [comment]
I copied to OAA materials, models and particles to 
I edited the lua and txt files. When I said I did not compiled I meant compiling the model on Hammer or something(editado)
do the Console say the material/file not found?

**Comment by carlosrpg** [comment]
The models, textures and particles are pre compiled by valve and I placed them on Game folder.
Opening the model or texture on valve editors will show that they are not editable (a lock icon on the side of the file) and the Content File: no file, that is because there is not original file only the compiled one.
 I'm adding a list of all files in my oaa folder (results of cmd call: dir /a-d /b /s > Files.txt) and the screen with the boss on the oaa map.

[Files.txt](https://github.com/OpenAngelArena/oaa/files/1523495/Files.txt)
image
image

**Comment by carlosrpg** [comment]
Chris, did it work? Let me know if we need anything to merge it to the game branch

**Comment by chrisinajar** [comment]
image
I don't seem to have most of those vmdl files

**Comment by carlosrpg** [comment]
Yep. I forgot to git unignore the model/creeps/ogre_1 . I'll do that when I get home today.
I'll copy my branch, delete everything and get all from git again to make sure it works

**Comment by carlosrpg** [comment]
Chris, it is all there already...
I deleted all my oaa data and downloaded the head again from git and it works just fine.
the model folder is game\models\creeps\ogre_1 and you can browser them from the github web.

I tested it in game and it works. Please Try again.

**Pull request merged**